src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm

```diff
@@ -1567,13 +1607,75 @@ machine(MachineType:Directory, "AMD Baseline protocol")
     ptl_popTriggerQueue;
   }
 
-  transition({BDR_PM, BDW_PM, BS_PM, BM_PM, B_PM, BDR_Pm, BDW_Pm, BS_Pm, BM_Pm, B_Pm, BP}, CPUPrbResp) {
+  transition({BDR_PM, BDW_PM, BS_PM, BM_PM, B_PM, BDR_Pm, BDW_Pm, BS_Pm, BM_Pm, B_Pm, BP, BL2}, CPUPrbResp) {
+    y_writeProbeDataToTBE;
+    x_decrementAcks;
+    o_checkForCompletion;
+    pr_popResponseQueue;
+  }
+
+  transition({BDR_PM, BDW_PM, BS_PM, BM_PM, B_PM, BDR_Pm, BDW_Pm, BS_Pm, BM_Pm, B_Pm, BP}, CPUPrbRespWB) {
     y_writeProbeDataToTBE;
     x_decrementAcks;
     o_checkForCompletion;
     pr_popResponseQueue;
   }
 
+  transition(BL2, L3Hit, BL2_Pm) {
+    ptl_popTriggerQueue;
+  }
+
+  transition({BL2, BL2_Pm}, CPUPrbRespWB, BL2_Pm) {
+    // Blocked on L2 and waiting for probes
+    yw_writeProbeDataToTBEWB;
+    x_decrementAcks;
+    o_checkForCompletion;
+    pr_popResponseQueue;
+  }
+
+  transition(BL2_Pm, CPUPrbResp) {
+    // Blocked on L2 probes, got the memory
+    x_decrementAcks;
+    o_checkForCompletion;
+    pr_popResponseQueue;
+  }
+
+  transition(BL2, ProbeAcksComplete, BL2_M) {
+    // We probed all the TCC dirs but didn't find the memory
+    // Send out memory request
+    // Transition to waiting on memory
+    pt_popTriggerQueue;
+  }
+
+  transition(BL2_Pm, ProbeAcksComplete, U) {
+    // We were waiting for all probes to come back now that they have we can unblock
+    // Send WBAck back to TCC
+    dd_sendResponseDmaData;
+    wada_wakeUpAllDependentsAddr;
+    dt_deallocateTBE;
+    pd_popDmaRequestQueue;
+    pt_popTriggerQueue;
+  }
+
+  transition(BL2, MemData, BL2_Pm) {
+    mt_writeMemDataToTBE;
+    pm_popMemQueue;
+  }
+
+  transition({BL2_Pm, U}, MemData) {
+    pm_popMemQueue;
+  }
```

Comment:
This seems suspect. Is there some case where duplicate memory requests get sent and we ignore their responses?

Reply:
The protocol changes are intended to work as follows when the cache line receives a DMAReadWB event trigger (only triggered if L2 is WB):

- The cache line moves from U to BL2. It checks L3 cache (if size is non-zero), probes L2 caches, and sends out a request to memory simultaneously.
- If it is a hit at L3 (line 1625) or if L2 sends back data (line 1630), the line transitions to BL2_Pm (got memory but waiting for probes to complete). Any data from memory is now ignored (line 1666)
- If it is not a hit at L3 and if L2 sends probe ack without data (line 1610), the cache waits for data from memory. If data from memory arrives before ProbeAcksComplete, the transition at line 1660 is called first and the transition at line 1650 is called when all probes are received
- If it is not a hit at L3 and if none of the L2s send data but send ProbeAcksComplete before memory request returns, the transition at line 1643 is called and the line moves to BL2_M state (waiting for mem). Then, when data from memory arrives, the transition at line 1669 is called.

There are four possible states for the cacheline when mem data returns: if it is in BL2 or BL2_M, the data is written to TBE and any dirty data in L2/L3 overwrites the TBE later. If it already received dirty data (in BL2_Pm) the mem data is ignored. If the line is in U (already received dirty data and all acks) the mem data is ignored.

I am unable to think of any other case. Please let me know if you can think of any, and I will add that too!

Reply:
@abmerop just checking if this addresses your concern?

Reply:
Thanks for the explanation!  It should then be impossible to get MemData in U state, correct?  This is the transition that concerns me.  MemData in BL2_Pm state makes sense.  If that is true, then the change would be `{BL2_Pm, U}` -> `BL2_Pm`.

The concern is it might be hiding some other bugs (potentially entirely unrelated to this PR) which go to U state when it should be waiting for memory.

Reply:
Sorry to belabor this, but I am still missing something.  In the case you describe wouldn't be in `BL2_M`? I.e., the state which is "got probes, waiting for memory"?

Reply:
The line would be in BL2_M only if none of the probes return dirty data. If a probe did, then the line moves to BL2_Pm.

There are two paths the sequence of transitions could take. When the state is in BL2, it is waiting for both probes from caches and response from memory. If the cache probes return dirty data, the line transitions to BL2_Pm. From here, either all the cache probes could return first (BL2_Pm->U) or memory could return first (remains in BL2_PM and moves to U once other probes return). If none of the cache probes returns dirty data, the line could move to BL2_M once all probes return but MemData hasn't or move to BL2_Pm if MemData comes first (BL2_Pm->U once probes return).

I will add more comments, change the state descriptions, and use a different name for the CPUPrbRespWB event (change it to CPUPrbRespDirty)